File: templates/problems/07_trees/12_kth_smallest_bst/typescript/solution.ts
import { TreeNode } from "./helpers";

export function solveKthSmallest(root: TreeNode | null, k: number): number {
    let count = 0;
    let result = 0;

    function inorder(node: TreeNode | null): void {
        if (node === null || count >= k) {
            return;
        }
        inorder(node.left);
        count++;
        if (count === k) {
            result = node.val;
            return;
        }
        inorder(node.right);
    }

    inorder(root);
    return result;
}

File: templates/problems/07_trees/12_kth_smallest_bst/python/solution.py
from helpers import TreeNode, Optional


def solve_kth_smallest(root: Optional[TreeNode], k: int) -> int:
    """Return the kth smallest value in a BST.

    Time: O(h + k), Space: O(h).
    """
    count = 0
    result = 0

    def inorder(node: Optional[TreeNode]) -> None:
        nonlocal count, result
        if node is None or count >= k:
            return
        inorder(node.left)
        count += 1
        if count == k:
            result = node.val
            return
        inorder(node.right)

    inorder(root)
    return result
